### DISCUSSION TITLE
Navigation Standards

### DISCUSSION THREAD

**Discussion by odudex** [selfcustody/krux#200] [General]
I've created some different ways to change view modes for SeedQRs, depending on the device. I could be confusing for some, so suggestions the help standardize navigation and make it more intuitive are welcome.
Currently users can swipe sideways to change view modes, swipe up/down to leave, press enter/touch will leave on standard view, but will increment line/region when at this view modes. Buttons page, page_prev and encoder rotation will change view modes.
It could be done differently by adding a submenu to make user choose mode before he sees the QR, page, page prev could increment lines/regions when possible and enter/touch would leave. Swipe between modes could still be present in touch devices.

**Comment by tadeubas** [comment]
I've noted that Krux makes use of this **patter for the UI**:
- `PAGE/PAGE_PREV/ENTER/TOUCH` is used to **advance or exit** in most of the screens, most exceptions are the menu screens and any screen that have the "Back" or "Esc" buttons (when I say TOUCH I mean when the user just clicks on the screen).

Some screens that work with **QRCode** don't accept `PAGE/PAGE_PREV` to advance or exit, these are:
- "Home > Sign > PSBT or Message":
  - After using camera to read a QRCode, the QRCodes presented by Krux on screen **only accepts** `ENTER/TOUCH` in order to **advance or exit**.
- "Home > Scan Address":
  - After using camera to read a QRCode, the QRCode presented by Krux on screen **only accepts** `ENTER/TOUCH` in order to **advance or exit** (Don't know why it presents a QRCode, because we can't compare with the one we just read, softwares uses different libs and create different QRCodes with the same content).
 - "Home > Wallet":
  - After loading a wallet reading a QRCode, when clicked again in this option, Krux presents a QRCode on screen that **only accepts** `ENTER/TOUCH` in order to **advance or exit**.
-  "Home > Extended Public Key":
  - This screen **only accepts** `ENTER/TOUCH` in order to **advance or exit** when presenting **QRCodes**.
- "Home > Mnemonic > Plaintext QR":
  - This screen **only accepts** `ENTER/TOUCH` in order to **advance or exit** because it shows a **QRCodes**.
  
After this brief overview of the current **Krux pattern on UI**, @odudex created new screens with different types of **QRCodes** like `SeedQR` and `Compact SeedQR`. These screens created new ways to interact with `PAGE/PAGE_PREV/ENTER` accepting only `ENTER/TOUCH` to **exit on specific screens** presented, and `PAGE/PAGE_PREV` to move around all screens (rotating forward on the screens, but never exiting).

First of all I would like to address the "traditional" Krux behaviour on QRCodes... it doesn't make sense to me that "Home > Extended Public Key" shows some information in text that accepts `PAGE/PAGE_PREV/ENTER/TOUCH` to advance but the QRCode presented just accepts `ENTER/TOUCH` to advance. I think that QRCodes doesn't make the screen special to just accept `ENTER/TOUCH` in order to **advance or exit**. I would make all screens that presents QRCode, to also accept `PAGE/PAGE_PREV` in order to **advance or exit**

@odudex commented that we can separate the various screens presented at `SeedQR` and `Compact SeedQR` in a sub-menu, presented just after clicking on on of these two items. The sub-menu could have: `Show QRCode`, `Matrix above QRCode`, `Line by line QRCode`, and so on... this would respect the current Krux UI we already have and allow the Krux pattern to continue. If we opt for this, @odudex expressed the idea to allow an exception on Amigo devices. The idea is that they would not need the sub-menu but instead continue using the screen is (an exception to the Krux UI pattern).

I know this is long... but it was necessary to highlight all of these cases to show what Krux "tells" the user on how to interact and what to expect... I am more in favor of a pattern than no pattern and let the user "guess" or have to tell the user how to interact with the device on that specific screen... my opinion is to allow exeptions when it is clear that the screen can't operate properly following the current Krux patter, or maybe we should change the UI pattern on all screens (we can allow user to set some kind of UI preference in settings too!).

**Comment by ghost** [comment]
I definitely agree that we should codify some of the UI/UX patterns and try to standardize them as much as possible. In general, I prefer minimalism and simplicity, which includes not explaining things to the users in the UI as much as possible. In my opinion, that's what docs and tutorials are for, not the UI itself. **But** to facilitate this, the UI/UX needs to be self-evident: intuitive, natural, and consistent.

> First of all I would like to address the "traditional" Krux behaviour on QRCodes... it doesn't make sense to me that "Home > Extended Public Key" shows some information in text that accepts PAGE/PAGE_PREV/ENTER/TOUCH to advance but the QRCode presented just accepts ENTER/TOUCH to advance. I think that QRCodes doesn't make the screen special to just accept ENTER/TOUCH in order to advance or exit. I would make all screens that presents QRCode, to also accept PAGE/PAGE_PREV in order to advance or exit

Ideally this would be the case, yes, but the reason QR codes were different is that you used to (or maybe still can) be able to page through animated QR codes to manually show specific QRs. So, out of necessity, QR code pages had to be behave differently. I'm not sure if this is still the case, though.

> @odudex commented that we can separate the various screens presented at SeedQR and Compact SeedQR in a sub-menu, presented just after clicking on on of these two items. The sub-menu could have: Show QRCode, Matrix above QRCode, Line by line QRCode, and so on... this would respect the current Krux UI we already have and allow the Krux pattern to continue. 

I think this is reasonable.

> my opinion is to allow exeptions when it is clear that the screen can't operate properly following the current Krux patter, or maybe we should change the UI pattern on all screens (we can allow user to set some kind of UI preference in settings too!).

Yes, that's why it's important when we add new features to keep in mind how they'll operate on the various devices, inputs, and form factors. In general, I think we should aim for uniformity of experience and not have features that only work on some devices. Exceptions can be made of course.

**Comment by tadeubas** [comment]
The camera QRCode capture is another screen that behaves different from "Krux Pattern" on M5stick and Amigo. ENTER activates flash and PAGE enable anti-glare mode, so on those devices you can only exit with PAGE_PREV or TOUCH

**Comment by tadeubas** [comment]
@jreesun I also noted that when a user set a Language locale different from the one that was set, the user automatically exits Settings screen and sees the Login page. This is a strange behavior because no other setting do the same when changed.

The code that does this:
https://github.com/selfcustody/krux/blob/dev-cleanup/src/krux/pages/login.py#L676

**Comment by tadeubas** [comment]
@odudex Added brightness control (using `BUTTON_PAGE`) on the screen for the animated QR codes (like the ones presented for signed PSBT)
